includes/embeds/class-amp-youtube-embed-handler.php

```diff
@@ -49,22 +50,93 @@ public function __construct( $args = [] ) {
 	 * Register embed.
 	 */
 	public function register_embed() {
-		wp_embed_register_handler( 'amp-youtube', self::URL_PATTERN, [ $this, 'oembed' ], -1 );
-		add_shortcode( 'youtube', [ $this, 'shortcode' ] );
+		add_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10, 2 );
+		add_shortcode( 'youtube', [ $this, 'shortcode' ] ); // @todo Deprecated. See <https://github.com/ampproject/amp-wp/issues/3309>.
 		add_filter( 'wp_video_shortcode_override', [ $this, 'video_override' ], 10, 2 );
 	}
 
 	/**
 	 * Unregister embed.
 	 */
 	public function unregister_embed() {
-		wp_embed_unregister_handler( 'amp-youtube', -1 );
-		remove_shortcode( 'youtube' );
+		remove_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10 );
+		remove_shortcode( 'youtube' ); // @todo Deprecated. See <https://github.com/ampproject/amp-wp/issues/3309>.
+	}
+
+	/**
+	 * Filter oEmbed HTML for YouTube to convert to AMP.
+	 *
+	 * @param string $cache Cache for oEmbed.
+	 * @param string $url   Embed URL.
+	 * @return string Embed.
+	 */
+	public function filter_embed_oembed_html( $cache, $url ) {
+		$host = wp_parse_url( $url, PHP_URL_HOST );
+		if ( ! in_array( $host, [ 'youtu.be', 'youtube.com', 'www.youtube.com' ], true ) ) {
+			return $cache;
+		}
+
+		$id = $this->get_video_id_from_url( $url );
+		if ( ! $id ) {
+			return $cache;
+		}
+
+		$props = $this->parse_props( $cache, $url, $id );
+
+		$props['video_id'] = $id;
+		return $this->render( $props, $url );
+	}
+
+	/**
+	 * Parse AMP component from iframe.
+	 *
+	 * @param string $html     HTML.
+	 * @param string $url      Embed URL, for fallback purposes.
+	 * @param string $video_id YouTube video ID.
+	 * @return array Props for rendering the component.
+	 */
+	private function parse_props( $html, $url, $video_id ) {
+		$props = [];
+
+		if ( preg_match( '#<iframe[^>]*?title="(?P<title>[^"]+)"#s', $html, $matches ) ) {
+			$props['title'] = $matches['title'];
+		}
+
+		$img = AMP_HTML_Utils::build_tag(
+			'img',
+			[
+				'src'             => esc_url_raw( sprintf( 'https://i.ytimg.com/vi/%s/hqdefault.jpg', $video_id ) ),
+				'alt'             => isset( $props['title'] ) ? $props['title'] : '',
+				'data-amp-layout' => 'fill',
+			]
+		);
+
+		$props['placeholder'] = AMP_HTML_Utils::build_tag(
+			'a',
+			[
+				'placeholder' => '',
+				'href'        => esc_url_raw( $url ),
+			],
+			$img
+		);
+
+		if ( preg_match( '#<iframe[^>]*?height="(?P<height>\d+)"#s', $html, $matches ) ) {
```

Comment:
Instead of having three separate regexes here that parse the iframe, it would be preferable to build a single one that parses all three attributes in one go.

See https://regex101.com/r/zErkQE/1 for an example of how to do that (with attributes in arbitrary order).

Reply:
That's a much better solution, thanks @schlessera :pray:

Reply:
I'm impressed. I didn't realize this could be done with a single pattern.

Reply:
@pierlon @schlessera I've applied the same change to the Soundcloud embed handler, and extracted the common logic to the base class. Please take a look.